File: 240410/디버깅/debugging.py
ans=4
n,m,h=map(int,input().split())
ladder=[[0 for _ in range(n+1)] for _ in range(h+1)]
def dfs(depth,last):
    global ans,n,m,h,ladder
    if(depth==ans):
        return
    p=True
    for i in range(1,n+1):
        num=i
        for j in range(1,h+1):
            if (ladder[j][num-1]):
                num-=1
            elif (ladder[j][num]):
                num+=1
        if(num!=i):
            p=False
            break
    if(p):
        ans=min(ans,depth)
        return
    if(depth==3):
        return;
    for i in range(last,h+1):
        for j in range(1,n):
            if (ladder[i][j]==0 and ladder[i][j-1]==0):
                ladder[i][j]=1
                dfs(depth+1,i)
                ladder[i][j]=0
for i in range(m):
    a,b=map(int,input().split())
    ladder[a][b]=1
dfs(0,1)
print(-1 if ans==4 else ans)

File: 240410/디버깅/debugging.cpp
#include <iostream>
using namespace std;
int ladder[31][11];
int n, m, h, ans = 4;
void dfs(int depth,int last)
{
    if (depth == ans)return;
    bool possible = true;
    for (int i = 1; i <= n; i++)
    {
        int num = i;
        for (int j = 1; j <= h; j++)
        {
            if (ladder[j][num - 1])num--;
            else if (ladder[j][num])num++;
        }
        if (num != i)
        {
            possible = false;
            break;
        }
    }
    if (possible)
    {
        ans = min(ans, depth);
        return;
    }
    if (depth == 3)return;
    for (int i = last; i <= h; i++)
    {
        for (int j = 1; j < n; j++)
        {
            if (!ladder[i][j] && !ladder[i][j - 1])
            {
                ladder[i][j] = 1;
                dfs(depth + 1,i);
                ladder[i][j] = 0;
            }
        }
    }
}
int main() {
    ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);
    cin >> n >> m >> h;
    for (int i = 0; i < m; i++)
    {
        int a, b;
        cin >> a >> b;
        ladder[a][b] = 1;
    }
    dfs(0,1);
    if (ans == 4)cout << -1;
    else cout << ans;
    return 0;
}